src/db.ts

```diff
@@ -0,0 +1,354 @@
+    var results = _.filter(sessions, function (item) {
+      return item.sessionId.indexOf(row.session_id) > -1;
+    });
+    if (results.length > 0) return;
+    sessions.push(
+      clean({
+        mediaType: row.media_type,
+        sessionId: row.session_id,
+        direction: row.direction,
+        provider: row.provider,
+        requestedRoutings: row.requested_routings ? row.requested_routings.split(',') : null,
+        usedRouting: row.used_routing,
+        selectedAgentId: row.selected_agent_id,
+        peerId: row.peer_id,
+        remote: row.remote,
+        segments: generateSegments(rows, row.session_id),
+        metrics: generateMetrics(rows, row.session_id),
+      })
+    );
+  });
+  return sessions;
+}
+
+// Generates segment objects from database rows
+function generateSegments(rows: any[], sessionId: string) {
+  var segments = [];
+  var segmentIds = [];
+  _.forEach(rows, function (row) {
+    if (row.session_id !== sessionId) return;
+    const segmentId = `${row.session_id}_${row.segment_start}_${row.segment_end}`;
+    if (segmentIds.includes(segmentId)) return;
+    segmentIds.push(segmentId);
+    segments.push(
+      clean({
+        segmentStart: row.segment_start,
+        segmentEnd: row.segment_end,
+        queueId: row.queue_id,
+        disconnectType: row.disconnect_type,
+        segmentType: row.segment_type,
+        conference: row.conference,
+      })
+    );
+  });
+  return segments;
+}
+
+// Generates metric objects from database rows
+function generateMetrics(rows: any[], sessionId: string) {
+  var metrics = [];
+  var metricIds = [];
+  _.forEach(rows, function (row) {
+    if (row.session_id !== sessionId) return;
+    const metricId = `${row.session_id}_${row.name}`;
+    if (metricIds.includes(metricId)) return;
+    metricIds.push(metricId);
+    metrics.push(
+      clean({
+        name: row.name,
+        value: row.value,
+        emitDate: row.emit_date,
+      })
+    );
+  });
+  return metrics;
+}
+
+/*
+  Exported functions
+*/
+
+export const createTables = async () => {
+  await createTable(CREATE_CONVERSATIONS_TABLE_QUERY, CONVERSATIONS);
+  await createTable(CREATE_PARTICIPANTS_TABLE_QUERY, PARTICIPANTS);
+  await createTable(CREATE_SESSIONS_TABLE_QUERY, SESSIONS);
+  await createTable(CREATE_SEGMENTS_TABLE_QUERY, SEGMENTS);
+  await createTable(CREATE_METRICS_TABLE_QUERY, METRICS);
+};
+
+export const dropTables = async () => {
+  await dropTable(METRICS, true);
+  await dropTable(SEGMENTS, true);
+  await dropTable(SESSIONS, true);
+  await dropTable(PARTICIPANTS, true);
+  await dropTable(CONVERSATIONS, true);
+};
+
+export const fetchConversation = async (conversationId: string) => {
+  const fetched = await fetchData(FETCH_CONVERSATION_QUERY, [conversationId]);
+  if (!fetched) return null;
+
+  var conversation = generateConversation(fetched);
+  return clean(conversation);
+};
+
+export const insertConversation = async (conversation: any) => {
```

Comment:
This should use transactions so it can roll back if any of the operations fail. The gist is that you begin a transaction, execute commands, then commit or rollback the transaction. https://www.postgresql.org/docs/current/tutorial-transactions.html

Reply:
You should also have the script write the bad record out to the log (or at least the id) if an error is encountered.

Reply:
I'll add a log of the conversation ID when an error happens.